### PR TITLE
WIP: Custom error types

### PR DESCRIPTION
**Pull request by rustonaut**
Adds the possibility of having custom error types (instead of `BoxedError`) in `FromRequest`, `Guard` and `FromBody`.

The custom error type is specified using a attribute like `#[error(MyErrorType)]` on the type where `FromRequest` is derived on (the route type).

Any `Guard`, `#[body]` or `#[forward]` in the route type can have a different custom error type. A all of them need to implement `Into` so that they can be converted to the rout's error type. (In practice you will likely implement `From` for all required error types for your custom error type.

### CODE REVIEW

**Comment by rustonaut** [comment]
It seems that it's not compatible with rust `1.32` for some reason... (but works on `1.36`)

File: src/service.rs [hunk 1]
```diff
@@ -392,22 +390,21 @@ where
...
-                crate::blocking(move || Ok(handler(route, req)))
+                crate::blocking(move || (handler)(route, req)).map_err(FromRequestError::Custom)
             })
             .map(move |response| {
                 if is_head {
                     // Responses to HEAD requests must have an empty body
                     response.map(|_| Body::empty())
                 } else {
-                    response
+                    response.map(|body| body.into())
```
Comment (rustonaut) [inline src/service.rs]: Wups, I need to remove that.

File: src/service.rs [hunk 2]
```diff
@@ -77,22 +76,22 @@ use std::sync::Arc;
 /// [`hyperdrive::Error`]: ../struct.Error.html
 pub struct AsyncService<H, R, F>
 where
-    H: Fn(R, Arc<Request<()>>) -> F + Send + Sync + 'static,
+    H: Fn(R, Arc<Request<()>>) -> F + Send + Sync + Clone + 'static,
```
Comment (jonas-schievink) [inline src/service.rs]: Where did the `Clone` requirement come from?

File: src/lib.rs [hunk 3]
```diff
@@ -227,15 +232,23 @@ pub type BoxedError = Box<dyn std::error::Error + Send + Sync>;
 /// JSON body.
 ///
 /// The generated `FromRequest` implementation will always use
-/// [`DefaultFuture<Self, BoxedError>`][`DefaultFuture`] as the associated
-/// `Result` type.
+/// [`DefaultFuture<Self, Self::Error>`][`DefaultFuture`] as the associated
+/// `Result` type and [`NoCustomError`] as `Error` type.
+///
+/// **Be aware that any custom error is a `hyper::Service` error and as such
+/// isn't handled by hyper by dropping the connection. To have custom error
```
Comment (jonas-schievink) [inline src/lib.rs]:
```suggestion
/// is handled by hyper by dropping the connection. To have custom error
```

File: src/lib.rs [hunk 4]
```diff
@@ -227,15 +232,23 @@ pub type BoxedError = Box<dyn std::error::Error + Send + Sync>;
...
+/// handling you have to wrap the created `hyper::Service` (e.g. wrap the
+/// result of [`SyncService::new`]**
```
Comment (jonas-schievink) [inline src/lib.rs]:
```suggestion
/// result of [`SyncService::new`])**
```

File: src/lib.rs [hunk 5]
```diff
@@ -227,15 +232,23 @@ pub type BoxedError = Box<dyn std::error::Error + Send + Sync>;
...
 ///
 /// Note that the generated implementation will make use of `.and_then()` to
 /// chain asynchronous operations instead of running them in parallel using
 /// `join_all`. This is because it simplifies the code and doesn't require
 /// making use of boxed futures everywhere in the generated code. Multiple
 /// requests will still be handled in parallel, so this should not negatively
-/// affect performance.
+/// affect performance. Note that the only operations which could be executed
+/// in parallel are the creations of all `Guard` instances (and maybe the
+/// `FromBody` instance). So for many routes this is a irrelevant implementation
```
Comment (jonas-schievink) [inline src/lib.rs]:
```suggestion
/// `FromBody` instance), so for many routes this is an irrelevant implementation
```

File: src/service.rs (hunk 2 above)
Comment (rustonaut) [reply src/service.rs]: Hm, it's a left over I will remove it (I had to go through some iterations as I did run into some rust limitations).

File: src/error.rs [hunk 6]
```diff
@@ -1,41 +1,255 @@
...
+    /// [`BuildInError`] used by generated [`FromRequest`] implementations.
+    ///
+    /// [`SyncService`] and [`AsyncService`] will automatically convert this
+    /// error to a appropriate HTTP response.
+    ///
+    /// [`BuildInError`]: struct.BuildInError.html
+    /// [`FromRequest`]: trait.FromRequest.html
+    /// [`SyncService`]: service/struct.SyncService.html
+    /// [`AsyncService`]: service/struct.AsyncService.html
+    BuildIn(BuildInError),
```
Comment (jonas-schievink) [inline src/error.rs]:
```suggestion
    BuiltIn(BuiltInError),
```

File: src/error.rs [hunk 7]
```diff
@@ -1,41 +1,255 @@
...
+    where
+        hyper::Error: Into<E>,
+    {
+        FromRequestError::Custom(h.into())
+    }
+
+    /// If this error is a [`FromRequestError::BuildIn`] return a reference to it.
+    ///
+    /// [`FromRequestError::BuildIn`]: struct.FromRequestError.html#variant.BuildIn
+    pub fn as_build_in(&self) -> Option<&BuildInError> {
```
Comment (jonas-schievink) [inline src/error.rs]:
```suggestion
    pub fn as_built_in(&self) -> Option<&BuiltInError> {
```

File: src/error.rs [hunk 8]
```diff
@@ -1,41 +1,255 @@
...
+    fn from(e: E) -> Self {
+        FromRequestError::Custom(e)
+    }
+}
 
 /// The error type used by this library.
 #[derive(Debug)]
-pub struct Error {
-    kind: ErrorKind,
+pub struct BuildInError {
```
Comment (jonas-schievink) [inline src/error.rs]:
```suggestion
pub struct BuiltInError {
```

File: src/error.rs [hunk 9]
```diff
@@ -1,41 +1,255 @@
...
+    kind: BuildInErrorKind,
```
Comment (jonas-schievink) [inline src/error.rs]:
```suggestion
    kind: BuiltInErrorKind,
```

File: src/lib.rs [hunk 10]
```diff
@@ -517,17 +534,49 @@ pub trait FromRequest: Sized {
     /// [`RequestContext`]: trait.RequestContext.html
     type Context: RequestContext;
 
+    /// Custom error used for this `FromRequest` implementation.
+    ///
+    /// This allows returning custom error types from [`Guard`] and [`FromBody`]
+    /// implementations as well as from the rout handling function.
```
Comment (jonas-schievink) [inline src/lib.rs]:
```suggestion
    /// implementations as well as from the route handling function.
```

File: src/error.rs [hunk 11]
```diff
@@ -1,41 +1,255 @@
...
+            BuildIn(err) => Some(err),
+        }
+    }
+}
+
+impl<E> Display for FromRequestError<E>
+where
+    E: Display + 'static,
+{
+    fn fmt(&self, fter: &mut fmt::Formatter<'_>) -> fmt::Result {
```
Comment (jonas-schievink) [inline src/error.rs]: The rest of the codebase calls it `f`, so perhaps for consistency:

```suggestion
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

File: src/error.rs [hunk 12]
```diff
@@ -1,41 +1,255 @@
...
+        use self::FromRequestError::*;
+        match self {
+            Custom(err) => Display::fmt(err, fter),
+            BuildIn(err) => Display::fmt(err, fter),
```
Comment (jonas-schievink) [inline src/error.rs]: This should work too, and is slightly shorter:

```suggestion
            BuildIn(err) => err.fmt(fter),
```

**Comment by rustonaut** [comment]
The think is before we already had 2 error types: Such which where handled by hyperdrive internally and such which where passed through to hyper. This patch does "just" expose the later one and makes them type safe.

While we could probably always box any hyperdrive error and other errors and then downcast this has multiple drawbacks:

1. No longer having the "right" error type in `hyper::service::Service` so any wrapper which handles a specific type of error now has to handle error trait objects, which means an additional error case is introduced ("error handler receives custom error but can not handle custom error as it was the wrong custom error"). Worse in difference to the other error cases that one is nearly always a bug. 
2. Also wrapping error handlers which are generic over any error type T which implements some custom trait can no longer be used and need to be wrapped them self.
2. No longer having a clear distinction between errors returned by the service and errors handled internally by hyperdrives. Having this distinction is IMHO important as wrongly using non hyperdrive builtin errors can easily lead to a bug.
3. ~~no unnecessary downcasts~~ (doesn't really matter)

File: src/error.rs (hunk 12 above)
Comment (rustonaut) [reply src/error.rs]: It won't work if Debug is in scope as it then is ambiguous. So I got used to always use the UFS for Debug/Display forwarding.

**Comment by rustonaut** [comment]
Including the consideration to maybe add error handling directly to hyperdrive I might have found a alternative more ergonomic solution:

- Change `SynService` to return a result, maybe `E=BoxedError` or `E: Into<BoxedError>`
- We add a `#[error_handler(for ErrorType use function)` attribute to the `FromRequest` derive.
- The error handler function has a signature of `fn(context: Self::Context,  error:  ErrorType) -> Result<Response<Body>, BoxedError>` if it returns a `Err` the error is not handled and passed to hyper.
- There can be multiple of them for different types selection is done by trying to downcast in order.
- `#[error_handler(for _ use function)]` can be used as a fallback (which directly takes the `BoxedError` without downcast.
- There are defaults in place for the fallback and the `for hyper::Error` errors, the one for hyper errors
   does just forwards them the fallback one panics (indirectly logging the Bug of no proper error handling
   through the normal panic output/logging).

Once that is done we can extend it to support following:
- `#[panic_handler(function)]` (or we also provide a `#[panic_handler(for _ use func]`)
- This defaults to a handler producing a empty body 500 Internal Server Error.
- Have some way to disable it (e.g. if we go with `for TypeOr_ use func` we could go with
  `for _ use reraise` and if it's the only panic handler registered just not use `catch_unwind`

**Comment by jonas-schievink** [comment]
Hmm, yeah that's also a possibility. However it does seem like a *lot* of infrastructure to add to hyperdrive, and ideally I'd like to keep it as small and focused as possible, without adding too many features that would make it look more like an entire Web-Framework.

A more ergonomic way of wrapping a hyper `Service`/`NewService` could help with making panic-catching and custom error responses simpler. I will check whether https://github.com/hyperium/hyper/issues/1782 makes this any easier.

Perhaps we should start by mapping out the different use cases for everything around error-handling, panic-handling and custom errors in https://github.com/dac-gmbh/hyperdrive/issues/22 to get a clearer picture of what a minimal solution would look like?

**Pull request closed without merging**